python/app/input_widget.py

```diff
@@ -171,18 +171,27 @@ def execute(self):
 
         # exec the python code, redirecting any stdout to the ouptut signal.
         # also redirect stdin if need be
-        with nested(self._stdout_redirect, self._stdin_redirect):
-            try:
-                # use our copy of locals to allow persistence between executions
-                if eval_code:
+        if eval_code:
+            with nested(self._stdout_redirect, self._stdin_redirect):
+                try:
+                    # use our copy of locals to allow persistence between executions
                     results = eval(python_code, globals(), self._locals)
-                    self.results.emit(str(results))
+                except Exception:
+                    # oops, error encountered. write/redirect to the error signal
+                    with self._stderr_redirect as stderr:
```

Comment:
is it simpler to just say `self._stderr_redirect.write(self._format_exc())`?

Reply:
`self._stderr_redirect` is a custom context manager and `QObject` (so that it can emit signals) that handles temporarily redirecting `STDERR`. So it has to be called this way in order to do what it needs to do. :)